# tiwih: tools i wish(ed) i had

+ **fairegions**: (fai) generate equal length regions from an fai (for parallelization)
+ **combine_slivar_counts**: combine counts from [slivar](https://github.com/brentp/slivar) for original call and for compound-hets
+ **highcov**: (bam/cram) find high-coverage regions (~3 minutes per 30X cram)
+ **meandepth**: (bam/cram) quickly calculate meandepth by sampling (~2 seconds)
+ **samplename**: (bam/cram) print sample name from readgroup tag
+ **setsvalt**: (vcf/bcf) set the ALT allele in a manta VCF to left + N + right for paragraph genotyper
+ **setref**: (vcf/bcf) change ref allele from N to the actual allele.
+ **slivar_split_fam**: (vcf/bcf) split joint slivar files by family and inheritance mode
+ **sum_slivar_counts**: sum counts from [slivar](https://github.com/brentp/slivar) from same samples split across multiple files/regions
+ **vcf2bedpe** : convert and SV VCF to simple bedpe

